## About dataset

##The dataset is taken from the UCI Machine Learning Repository and is also present in sklearn's datasets module. It has 14 explanatory variables describing various aspects of residential homes in Boston, the challenge is to predict the median value of owner-occupied homes per $1000s.

In [17]:
import xgboost as xgb
from sklearn.metrics import mean_squared_error
import pandas as pd
import numpy as np

In [18]:
from sklearn.datasets import load_boston
boston = load_boston()

In [19]:
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [20]:
boston.data.shape

(506, 13)

In [21]:
boston.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [22]:
boston.DESCR

".. _boston_dataset:\n\nBoston house prices dataset\n---------------------------\n\n**Data Set Characteristics:**  \n\n    :Number of Instances: 506 \n\n    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.\n\n    :Attribute Information (in order):\n        - CRIM     per capita crime rate by town\n        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.\n        - INDUS    proportion of non-retail business acres per town\n        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)\n        - NOX      nitric oxides concentration (parts per 10 million)\n        - RM       average number of rooms per dwelling\n        - AGE      proportion of owner-occupied units built prior to 1940\n        - DIS      weighted distances to five Boston employment centres\n        - RAD      index of accessibility to radial highways\n        - TAX      full-value property-tax rate per $10,000

## Convert data into pandas dataframe

In [24]:

data = pd.DataFrame(boston.data)
data.columns = boston.feature_names

In [25]:
data.head()

CRIM    ZN  INDUS  CHAS    NOX  ...  RAD    TAX  PTRATIO       B  LSTAT
0  0.00632  18.0   2.31   0.0  0.538  ...  1.0  296.0     15.3  396.90   4.98
1  0.02731   0.0   7.07   0.0  0.469  ...  2.0  242.0     17.8  396.90   9.14
2  0.02729   0.0   7.07   0.0  0.469  ...  2.0  242.0     17.8  392.83   4.03
3  0.03237   0.0   2.18   0.0  0.458  ...  3.0  222.0     18.7  394.63   2.94
4  0.06905   0.0   2.18   0.0  0.458  ...  3.0  222.0     18.7  396.90   5.33

[5 rows x 13 columns]

In [26]:
data['PRICE'] = boston.target

In [27]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  PRICE    506 non-null    float64
dtypes: float64(14)
memory usage: 55.5 KB


In [28]:
data.describe()

CRIM          ZN       INDUS  ...           B       LSTAT       PRICE
count  506.000000  506.000000  506.000000  ...  506.000000  506.000000  506.000000
mean     3.613524   11.363636   11.136779  ...  356.674032   12.653063   22.532806
std      8.601545   23.322453    6.860353  ...   91.294864    7.141062    9.197104
min      0.006320    0.000000    0.460000  ...    0.320000    1.730000    5.000000
25%      0.082045    0.000000    5.190000  ...  375.377500    6.950000   17.025000
50%      0.256510    0.000000    9.690000  ...  391.440000   11.360000   21.200000
75%      3.677083   12.500000   18.100000  ...  396.225000   16.955000   25.000000
max     88.976200  100.000000   27.740000  ...  396.900000   37.970000   50.000000

[8 rows x 14 columns]

## Separate the target variable and rest of the variables using .iloc

In [29]:
X, y = data.iloc[:,:-1],data.iloc[:,-1]


##convert the dataset into an optimized data structure called dmatrix

In [31]:
data_dmatrix = xgb.DMatrix(data=X,label=y)

## split data into train and test

In [32]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)

In [33]:
xg_reg = xgb.XGBRegressor(objective ='reg:linear', colsample_bytree = 0.3, learning_rate = 0.1,
                max_depth = 5, alpha = 10, n_estimators = 10)

In [35]:
xg_reg.fit(X_train,y_train)

[05:43:51] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


XGBRegressor(alpha=10, base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=0.3, gamma=0,
             importance_type='gain', learning_rate=0.1, max_delta_step=0,
             max_depth=5, min_child_weight=1, missing=None, n_estimators=10,
             n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
             silent=None, subsample=1, verbosity=1)

In [36]:
preds = xg_reg.predict(X_test)

In [37]:
preds

array([ 8.531183 , 17.426376 , 24.354538 , 10.202603 , 17.86703  ,
       20.111704 , 17.825245 ,  7.0048833, 12.371497 , 18.518875 ,
       18.143133 , 14.239707 , 11.409273 , 17.309526 , 13.857576 ,
       14.114464 , 13.485846 , 26.225672 , 13.496281 , 10.878549 ,
       11.365604 , 12.844106 , 18.621658 , 22.8889   , 22.134113 ,
       13.578582 , 10.041029 , 14.692133 , 14.988752 , 13.096905 ,
       13.522167 , 19.658325 ,  7.2066092, 14.5406   , 15.21958  ,
       22.480894 , 16.636026 , 10.085008 , 12.709627 , 23.140484 ,
       16.660902 , 13.496281 ,  8.958481 , 24.830044 , 12.231392 ,
       17.061684 , 13.437938 , 13.822018 , 12.229067 , 10.875528 ,
       19.658325 , 16.804878 , 13.622031 ,  7.8421807, 14.484597 ,
       10.775787 , 11.339119 ,  7.332822 , 19.603703 ,  7.9418693,
       12.913391 , 13.190836 , 11.924444 , 11.435618 , 14.4757595,
       15.929399 , 16.288265 , 12.295227 , 15.335289 , 18.830666 ,
       13.095968 , 15.788654 , 12.999216 , 17.24653  , 10.5440

##Computing the rmse by invoking the mean_sqaured_error

In [39]:
rmse = np.sqrt(mean_squared_error(y_test, preds))

In [40]:
print("RMSE: %f" % (rmse))

RMSE: 10.449300


##Well, we can see that our RMSE for the price prediction came out to be around 10.8 per 1000$.